### PR TITLE
[GLUTEN-3521][CH] Fix substring start index can not be 0

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#3521)

## How was this patch tested?

TEST BY UT

### CODE REVIEW

File: cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    const ActionsDAG::Node * parse(
+    const substrait::Expression_ScalarFunction & substrait_func,
+    ActionsDAGPtr & actions_dag) const override
+    {
+        auto parsed_args = parseFunctionArguments(substrait_func, "", actions_dag);
+        if (parsed_args.size() != 2 && parsed_args.size() != 3)
+            throw Exception(DB::ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} requires two or three arguments", getName());
+        DB::DataTypePtr start_index_data_type = removeNullable(parsed_args[1]->result_type);
+        if (!DB::WhichDataType(start_index_data_type).isInt())
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp:49]: use isInteger make is simpler.

File: cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        */
+        auto * const_zero_node = addColumnToActionsDAG(actions_dag, start_index_data_type, Field(0));
+        auto * const_one_node = addColumnToActionsDAG(actions_dag, start_index_data_type, Field(1));
+        auto * equals_zero_node = toFunctionNode(actions_dag, "equals", {parsed_args[1], const_zero_node});
+        auto * index_plus_node = toFunctionNode(actions_dag, "plus", {parsed_args[1], const_one_node});
+        auto * if_node = toFunctionNode(actions_dag, "if", {equals_zero_node, index_plus_node, parsed_args[1]});
+        if (parsed_args.size() == 2)
+            return toFunctionNode(actions_dag, "substringUTF8", {parsed_args[0], if_node});
+        else
+            return toFunctionNode(actions_dag, "substringUTF8", {parsed_args[0], if_node, parsed_args[2]});
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp:66]: wrap convertNodeTypeIfNeeded

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp (hunk 1 above)
Comment (KevinyhZou) [reply cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp]: done

File: cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp (hunk 2 above)
Comment (KevinyhZou) [reply cpp-ch/local-engine/Parser/scalar_function_parser/substring.cpp]: done

**Review by binmahone — APPROVED** [review]

**Pull request merged**